src/Ethereum/Barz.cpp

```diff
@@ -85,4 +86,126 @@ Data getFormattedSignature(const Data& signature, const Data challenge, const Da
+    const Data& barzMsgHashData = parse_hex("0xb1bcb804a4a3a1af3ee7920d949bdfd417ea1b736c3552c8d6563a229a619100");
+
+    auto encodedDomainSeparatorData = Ethereum::ABI::Function::encodeParams(Ethereum::ABI::BaseParams {
+        std::make_shared<Ethereum::ABI::ProtoBytes32>(domainSeparatorTypeHashData),
+        std::make_shared<Ethereum::ABI::ProtoUInt256>(chainId),
+        std::make_shared<Ethereum::ABI::ProtoAddress>(barzAddress)
+    });
+    if (!encodedDomainSeparatorData.has_value()) {
+        return {};
+    }
+
+    Data domainSeparator = encodedDomainSeparatorData.value();
+    const Data domainSeparatorHash = Hash::keccak256(domainSeparator);
+
+    auto encodedRawMessageData = Ethereum::ABI::Function::encodeParams(Ethereum::ABI::BaseParams {
+        std::make_shared<Ethereum::ABI::ProtoBytes32>(barzMsgHashData),
+        std::make_shared<Ethereum::ABI::ProtoBytes32>(Hash::keccak256(msgHash)),
+    });
+
+    Data rawMessageData = encodedRawMessageData.value();
+
+    auto encodedMsg = Ethereum::ABI::Function::encodeParams(Ethereum::ABI::BaseParams {
+        std::make_shared<Ethereum::ABI::ProtoBytes32>(domainSeparatorHash),
+        std::make_shared<Ethereum::ABI::ProtoBytes32>(Hash::keccak256(rawMessageData))
+    });
+    auto encodedMsgData = "0x1901" + hex(encodedMsg.value());
+
+    Data finalEncodedMsgData = parse_hex(encodedMsgData);
+
+    const Data encodedMsgHash = Hash::keccak256(finalEncodedMsgData);
+
+    Data envelope;
+    append(envelope, encodedMsgHash);
+
+    return envelope;
+}
+
+// Function to encode the diamondCut function call using protobuf message as input
+Data getDiamondCutCode(const Proto::DiamondCutInput& input) {
+    Data encoded;
+
+    //    function diamondCut(
+    //        FacetCut[] calldata diamondCut,
+    //        address init,
+    //        bytes calldata _calldata // Note that Barz does not use the _calldata for initialization.
+    //    )
+    Data encodedSignature = parse_hex("1f931c1c"); // diamondCut() function selector
+    encoded.insert(encoded.end(), encodedSignature.begin(), encodedSignature.end());
+
+    // First argument Data Location `diamondCut`
+    Data dataLocation = parse_hex("60");
+    pad_left(dataLocation, 32);
+    append(encoded, dataLocation);
+
+    // Encode second Parameter `init`
+    Data initAddress = parse_hex(input.init_address());
+    pad_left(initAddress, 32);
+    append(encoded, initAddress);
+
+    // Third Argument Data location `_calldata`
+    auto callDataDataLocation = int(hex(encoded).size()) / 2;
+
+    Ethereum::ABI::ValueEncoder::encodeUInt256(input.facet_cuts_size(), encoded);
+
+    // Prepend the function selector for the diamondCut function
+    int instructChunk = 0;
+    int totalInstructChunk = 0;
+    int prevDataPosition = 0;
+    const auto encodingChunk = 32;
+    const auto bytesChunkLine = 5;
+    int chunkLocation;
+    Data dataPosition;
+    // Encode each FacetCut from the input
```

Comment:
Is this EIP2535? I think it's worth to move the implementation to Rust, as we have configured fuzz tests to tests such encodings, and in the end, we're trying to migrate existing code to Rust.
Do we have time for the refactoring? If so, I can explain where and how to implement the encoding, otherwise we'll consider moving it later.

Reply:
Yeah considering that this task is time-sensitive, I'll put the rust migration in the backlog to implement it when we have capacity. (Also, would be good if the whole Barz related code could be migrated to Rust)
For now, I think we can continue with the existing cpp code like other Barz related codes.